File: backend/app/api/spending.py
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.spending import PurchaseHistoryCreate, PurchaseHistoryRead, SpendingMetrics
from app.services.spending_service import SpendingService

router = APIRouter(prefix="/spending", tags=["spending"])


@router.get("/history", response_model=list[PurchaseHistoryRead])
async def get_history(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await SpendingService(db).list_for_user(current_user.id)


@router.post("/record", response_model=PurchaseHistoryRead, status_code=201)
async def record_purchase(
    data: PurchaseHistoryCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await SpendingService(db).record(current_user.id, data)


@router.get("/metrics", response_model=SpendingMetrics)
async def get_metrics(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await SpendingService(db).get_metrics(current_user.id)

File: frontend/src/api/spending.ts
import apiClient from './client';
import type { PurchaseHistory, SpendingMetrics, RecordPurchasePayload } from '../types';

export const getSpendingHistory = async (): Promise<PurchaseHistory[]> => {
  const response = await apiClient.get<PurchaseHistory[]>('/spending/history');
  return response.data;
};

export const recordPurchase = async (payload: RecordPurchasePayload): Promise<PurchaseHistory> => {
  const response = await apiClient.post<PurchaseHistory>('/spending/record', payload);
  return response.data;
};

export const getSpendingMetrics = async (): Promise<SpendingMetrics> => {
  const response = await apiClient.get<SpendingMetrics>('/spending/metrics');
  return response.data;
};
